### PR TITLE
Wip update to graal20 √

### PR DESCRIPTION
**Pull request by viktorklang**
DO NOT MERGE

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -355,13 +350,16 @@ def sharedNativeImageSettings(targetDir: File) = Seq(
...
   "-H:+ReportExceptionStackTraces",
+  "-H:-TraceClassInitialization", // if "+" prints out all trace information about class initialization
   "-H:-PrintUniverse", // if "+" prints out all classes which are included
   "-H:-NativeArchitecture", // if "+" Compiles the native image to customize to the local CPU arch
   "-H:Class=" + "io.cloudstate.proxy.CloudStateProxyMain",
   "--verbose",
   //"--no-server", // Uncomment to not use the native-image build server, to avoid potential cache problems with builds
-  //"--report-unsupported-elements-at-runtime", // Hopefully a self-explanatory flag
+  //"--debug-attach=5005", // Debugger makes a ton of sense to use to debug SubstrateVM
+  "--report-unsupported-elements-at-runtime", // Hopefully a self-explanatory flag
```
Comment (viktorklang) [inline build.sbt:360]: We should get rid of this when we can—right now it is needed to pass the native-image generation, but I would rather have things surfacing at build time.

File: proxy/cassandra/src/main/java/io/cloudstate/internal/svm/Substitutions.java [hunk 2]
```diff
@@ -15,16 +15,6 @@
   public static int REF_ELEMENT_SHIFT;
 }
 
-@TargetClass(className = "io.netty.util.internal.CleanerJava6")
```
Comment (viktorklang) [inline proxy/cassandra/src/main/java/io/cloudstate/internal/svm/Substitutions.java:18]: These are now done automatically by graalvm 20

File: proxy/core/src/main/java/io/cloudstate/internal/svm/Substitutions.java [hunk 3]
```diff
@@ -16,16 +16,6 @@
   public static int REF_ELEMENT_SHIFT;
 }
 
-@TargetClass(className = "io.grpc.netty.shaded.io.netty.util.internal.CleanerJava6")
```
Comment (viktorklang) [inline proxy/core/src/main/java/io/cloudstate/internal/svm/Substitutions.java:19]: These are now done automatically by graalvm 20

File: tck/src/it/resources/application.conf [hunk 4]
```diff
@@ -17,7 +17,7 @@ cloudstate-tck.combinations = [{
     //stop-command = ["docker", "stop", "cloudstate-proxy-tck"]
     // Switch to this to run the TCK with a native image that has been built for the local platform.
     // Make sure that the ENV JAVA_LIBRARY_PATH is set to the /lib dir of the JRE to load libraries from
-    //command = ["proxy/core/target/graalvm-native-image/cloudstate-proxy-core", "-Djava.library.path="${JAVA_LIBRARY_PATH}]
+    command = ["proxy/core/target/graalvm-native-image/cloudstate-proxy-core", "-Djava.library.path="${JAVA_HOME}"/lib", "-Dconfig.resource=in-memory.conf", "-Dcloudstate.proxy.dev-mode-enabled=true"]
```
Comment (viktorklang) [inline tck/src/it/resources/application.conf:20]: Making sure we run it:test against the generated native image.

**Comment by viktorklang** [comment]
Getting this weird error in the build:

```
java.lang.NoSuchFieldError: nullSession

	at sun.security.ssl.ClientHandshaker.getKickstartMessage(ClientHandshaker.java:1350)

	at sun.security.ssl.Handshaker.kickstart(Handshaker.java:1117)

	at sun.security.ssl.SSLEngineImpl.kickstartHandshake(SSLEngineImpl.java:736)

	at sun.security.ssl.SSLEngineImpl.beginHandshake(SSLEngineImpl.java:758)

	at akka.stream.impl.io.TLSActor.<init>(TLSActor.scala:162)

	at akka.stream.impl.io.TLSActor$.$anonfun$props$1(TLSActor.scala:38)

	at akka.actor.TypedCreatorFunctionConsumer.produce(IndirectActorProducer.scala:90)

	at akka.actor.Props.newActor(Props.scala:226)

	at akka.actor.ActorCell.newActor(ActorCell.scala:609)

	at akka.actor.ActorCell.create(ActorCell.scala:635)

	at akka.actor.ActorCell.invokeAll$1(ActorCell.scala:509)

	at akka.actor.ActorCell.systemInvoke(ActorCell.scala:531)

	at akka.dispatch.Mailbox.processAllSystemMessages(Mailbox.scala:294)

	at akka.dispatch.Mailbox.run(Mailbox.scala:229)

	at akka.dispatch.Mailbox.exec(Mailbox.scala:242)

	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)

	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)

	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)

	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)

2020-04-17 11:05:01.403 ERROR akka.actor.OneForOneStrategy - head of empty list

java.util.NoSuchElementException: head of empty list

	at scala.collection.immutable.Nil$.head(List.scala:430)

	at scala.collection.immutable.Nil$.head(List.scala:427)

	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:573)

	at akka.actor.ActorCell.invoke(ActorCell.scala:543)

	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:269)

	at akka.dispatch.Mailbox.run(Mailbox.scala:230)

	at akka.dispatch.Mailbox.exec(Mailbox.scala:242)

	at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)

	at java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1056)

	at java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1692)

	at java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)
```

Any ideas @srikanthops @jroper @hseeberger?

**Comment by ignasi35** [comment]
@viktorklang is this a nativeimage binary you are running? What GraalVM did you use? 
 GraalVM `20.0.0.r11` or `20.0.0.r8`?

**Comment by viktorklang** [comment]
@ignasi35 Good questions! This is what is in the `build.sbt`: `val GraalVersion = "20.0.0"`

File: build.sbt [hunk 5]
```diff
@@ -57,9 +57,8 @@ val AkkaPersistenceCassandraVersion = "0.102"
 val PrometheusClientVersion = "0.6.0"
 val ScalaTestVersion = "3.0.5"
 val ProtobufVersion = "3.9.0"
-val GraalVersion = "19.3.0"
-val DockerBaseImageVersion = "openjdk:8-jre-slim-buster"
-val svmGroupId = if (GraalVersion startsWith "19.2") "com.oracle.substratevm" else "org.graalvm.nativeimage"
+val GraalVersion = "20.0.0"
+val DockerBaseImageVersion = "openjdk:8-jre-buster"
```
Comment (viktorklang) [inline build.sbt:61]: @srikanthops We need to use the full, not the slim, jre to make sure that SSL etc works.

**Comment by ignasi35** [comment]
@viktorklang ah, this happens during the build? But it looks like a runtime error. Is that stacktrace thrown during a test? That's intriguing.

**Comment by viktorklang** [comment]
@ignasi35 No, I believe that the SSL-related things are loaded via the JAVA_LIBRARY_PATH, and in this case, the slim jre docker image probably do not have them, but rather stubs or truncated functionality, in which case at runtime this entire thing fails since it tries to read things that aren't there. That's what I suspect at least.

File: build.sbt [hunk 6]
```diff
@@ -57,9 +57,9 @@ val AkkaPersistenceCassandraVersion = "0.102"
 val PrometheusClientVersion = "0.6.0"
 val ScalaTestVersion = "3.0.5"
 val ProtobufVersion = "3.9.0"
-val GraalVersion = "19.3.0"
-val DockerBaseImageVersion = "openjdk:8-jre-slim-buster"
-val svmGroupId = if (GraalVersion startsWith "19.2") "com.oracle.substratevm" else "org.graalvm.nativeimage"
+val GraalVersion = "20.0.0"
+
+val DockerBaseImageVersion = "openjdk:8-jdk-slim-buster"
```
Comment (viktorklang) [inline build.sbt:62]: @srikanthops At least I got rid of the nullSession NoSuchFieldError, but I'm not sure this is the image we should be using. What do you advise?

File: build.sbt [hunk 7]
```diff
@@ -57,9 +57,9 @@ val AkkaPersistenceCassandraVersion = "0.102"
 val PrometheusClientVersion = "0.6.0"
 val ScalaTestVersion = "3.0.5"
 val ProtobufVersion = "3.9.0"
-val GraalVersion = "19.3.0"
-val DockerBaseImageVersion = "openjdk:8-jre-slim-buster"
-val svmGroupId = if (GraalVersion startsWith "19.2") "com.oracle.substratevm" else "org.graalvm.nativeimage"
+val GraalVersion = "20.0.0"
+
+val DockerBaseImageVersion = "adoptopenjdk/openjdk8:debian"
```
Comment (viktorklang) [inline build.sbt:62]: @srikanthops This image passes the tests, and should be based on Buster. Can you verify?

**Comment by viktorklang** [comment]
@hseeberger @jroper This is ready for review.

File: build.sbt (hunk 6 above)
Comment (srikanthops) [reply build.sbt]: [Travis CI Tests](https://travis-ci.com/github/cloudstateio/cloudstate/builds/152907832) passed with my [original pull request](https://github.com/cloudstateio/cloudstate/pull/207)

this image addresses/fixes a few vulnerabilities

**Comment by viktorklang** [comment]
@srikanthops check my proposed replacement image.

-- 
Cheers,
√

**Comment by viktorklang** [comment]
Already done in other branch

**Pull request closed without merging**